pkg/metrics/server/server.go

```diff
@@ -101,6 +101,10 @@ type Options struct {
 	// TLSOpts is used to allow configuring the TLS config used for the server.
 	// This also allows providing a certificate via GetCertificate.
 	TLSOpts []func(*tls.Config)
+
+	// ListenConfig contains options for listening to an address on the metric server.
+	// Note: This does not work with SecureServing=true
+	ListenConfig net.ListenConfig
```

Comment:
If this doesn't work for SecureServing=true do we want this to be an option?

Reply:
yeah I was concerned about this. I debated adding more complicity in the function to fail if both were set but decided I would allow that to be decided here. The reason I chose not to just support secure as well is that in that code path for it harder to inject.
```
...
tls.Listen("tcp", s.options.BindAddress, cfg)
...
```
Is used instead of net.Listen
```
package tls
...
func Listen(network, laddr string, config *Config) (net.Listener, error) {
	if config == nil || len(config.Certificates) == 0 &&
		config.GetCertificate == nil && config.GetConfigForClient == nil {
		return nil, errors.New("tls: neither Certificates, GetCertificate, nor GetConfigForClient set in Config")
	}
	l, err := net.Listen(network, laddr)
	if err != nil {
		return nil, err
	}
	return NewListener(l, config), nil
}
```
which means we probably would need this to do some of the responsibilities as it looks like it just uses net.Listen within it with a lot of other domain specific things. I couldn't see an easy way to inject the ListenConfig but maybe someone does.

Reply:
`tls.NewListener` you can pass a listener in (which can be a ListenConfig) right? 

Something like this: 

```go
l, err := s.options.ListenConfig.Listen(context.Background(), "tcp", s.options.BindAddress)
     if err != nil {
       return nil, err
   }
     return tls.NewListener(l, cfg), nil
```

I'm also not sure what this provides beyond the means you express. We may need a maintainer to discuss as well.  

However, I still believe we shouldn't break existing functionality so guarding across the `net.Listen` should happen if the option doesn't exist.

Reply:
Addressed

Reply:
> I'm also not sure what this provides beyond the means you express. We may need a maintainer to discuss as well.

Looking into this a little bit more, I know this behavior allows the sharing of the port with other "things" and can't confidently say this is something that controller-runtime is willing to support.  I am not a maintainer and not quite familiar with what the intended use case for this expanding beyond what is already there.  

I understand your use case but want to see what other implications this has with the metrics server and what fallout can exist with allowing the new configuration in.

Reply:
/cc @alvaroaleman

Reply:
I have addressed this it should support it for all cases now